### PR TITLE
PR: Separate pdb input from regular input

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes
A few part of the code now use the pdb command line to send messages to pdb while debugging. (For example, to ask pdb to check for incoming comm messages.) These change allow the frontend to separate between `input` calls that are done during debugging, and that should not recieve these messages, and pdb input. The detection is done by comparing the prompt to 'ipdb> '. 

All the direct calls to `input` or `write_to_stdin` are replaced by a new function `pdb_execute`, as this is what these calls are used for.

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 1]
```diff
@@ -122,7 +126,48 @@ def refresh_from_pdb(self, pdb_state):
...
+        self.add_to_pdb_history(self._pdb_line_num, line)
+
+        if line:
+            self._last_pdb_cmd = line
+
+        # This is the Spyder addition: add a %plot magic to display
+        # plots while debugging
+        if line.startswith('%plot '):
+            line = line.split()[-1]
+            line = "__spy_code__ = get_ipython().run_cell('%s')" % line
```
Comment (impact27) [inline spyder/plugins/ipythonconsole/widgets/debugging.py]: I am not sure what this code does. Should we keep it? What is it supposed to do?

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 2]
```diff
@@ -57,35 +57,39 @@ def __init__(self, *args, **kwargs):
...
+            # Print the string to the console
+            if not hidden:
+                # This emulates an user interaction
+                self._control.insert_text(line + '\n')
+                self._reading = False
+                self._append_before_prompt_cursor.setPosition(
+                    self._get_end_cursor().position())
+            self._input_ready = False
+            # Match ConsoleWidget.do_execute
+            self._executing = True
```
Comment (impact27) [inline spyder/plugins/ipythonconsole/widgets/debugging.py]: Should `self._executing = True` only be called when `not hidden`?

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 3]
```diff
@@ -133,59 +178,36 @@ def _handle_input_request(self, msg):
...
 
-            # Save history to browse it later
-            self._pdb_line_num += 1
-            self.add_to_pdb_history(self._pdb_line_num, line)
+        self._previous_prompt = (prompt, password)
 
-            if line:
-                self._last_pdb_cmd = line
+        if print_prompt:
+            self._highlighter.highlighting_on = True
```
Comment (impact27) [inline spyder/plugins/ipythonconsole/widgets/debugging.py]: Should the highlighter only be turned on for a pdb request? Here it is turned on for any request.

File: spyder/plugins/ipythonconsole/widgets/debugging.py (hunk 2 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/debugging.py]: I updated the PR to handle this more cleanly.

File: spyder/plugins/ipythonconsole/widgets/debugging.py [hunk 4]
```diff
@@ -122,7 +149,47 @@ def refresh_from_pdb(self, pdb_state):
...
+        # Fixes 2034
+        self.pdb_execute('continue')
+
     # ---- Private API (overrode by us) ----------------------------
+    def _readline_callback(self, line):
+        """Callback used when the user inputs text in stdin."""
+        if not self.is_waiting_pdb_input():
+            if self._input_ready:
+                # This is a regular input call
+                self._finalize_input_request()
```
Comment (impact27) [inline spyder/plugins/ipythonconsole/widgets/debugging.py]: This is maybe something that should be changed in qtconsole as well. I will open an issue there.

Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/debugging.py]: https://github.com/jupyter/qtconsole/issues/370

**Comment by impact27** [comment]
This will pass the tests when https://github.com/jupyter/qtconsole/pull/371 is merged.

**Comment by impact27** [comment]
Summary of work done:
 - Created a `pdb_execute` function to execute statments in pdb kernel while debugging. This replace the `stdin` and `input` calls.
 - Properly detect if a pdb command loop is running. Separate between pdb requests and general `input` requests. 
 - Remove debugging handling from `KernelComm` and place it in `debugging.py`.

File: requirements/conda.txt [hunk 5]
```diff
@@ -16,7 +16,7 @@ pyxdg
 pyzmq
 qdarkstyle >=2.7
 qtawesome >=0.5.7
-qtconsole >=4.5.2
+qtconsole >=4.5.5
```
Comment (ccordoba12) [inline requirements/conda.txt]: Please revert this change because it's not necessary given that we're pulling qtconsole from Github and we also need to wait until Anaconda releases packages for this version.

File: spyder/plugins/ipythonconsole/widgets/debugging.py (hunk 1 above)
Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/widgets/debugging.py]: It's used to generate inline plots while debugging by calling this `%plot` pseudo-magic I added to our debugger some time ago. Without this trick, plots can't be seen while debugging.

If you think this can implemented better, please leave it for a another PR.

File: spyder/plugins/ipythonconsole/widgets/debugging.py (hunk 3 above)
Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/widgets/debugging.py]: Yes, please only turn it on for Pdb requests.

**Comment by impact27** [comment]
I have no idea why `spyder/app/tests/test_mainwindow.py::test_lock_action` would fail. It doesn't fail on my computer.

**Comment by impact27** [comment]
re triggering the tests
